File: backend/app/models/project.py
from typing import Any

from sqlalchemy import Float, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.db.types import SpatialGeometry


class Project(Base):
    __tablename__ = 'projects'

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    country: Mapped[str] = mapped_column(String(120), nullable=False)
    status: Mapped[str] = mapped_column(String(30), default='draft', nullable=False)
    budget: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    description: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    created_by: Mapped[int | None] = mapped_column(ForeignKey('users.id'), nullable=True)

    sites: Mapped[list['Site']] = relationship(back_populates='project', cascade='all, delete-orphan')


class Site(Base):
    __tablename__ = 'sites'

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    project_id: Mapped[int] = mapped_column(ForeignKey('projects.id'), nullable=False)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    status: Mapped[str] = mapped_column(String(30), default='planned', nullable=False)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    note: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    geometry: Mapped[Any | None] = mapped_column(SpatialGeometry(), nullable=True)

    project: Mapped['Project'] = relationship(back_populates='sites')

File: frontend/src/types/project.ts
export type ProjectStatus = 'draft' | 'active' | 'paused' | 'completed'
export type SiteStatus = 'planned' | 'monitoring' | 'flagged' | 'resolved'

export type Project = {
  id: number
  name: string
  country: string
  status: ProjectStatus
  budget: number
  description?: string | null
  site_count: number
}

export type Site = {
  id: number
  project_id: number
  name: string
  status: SiteStatus
  latitude?: number | null
  longitude?: number | null
  note?: string | null
}

export type Geometry = {
  type: string
  coordinates: unknown[]
}

export type SiteDetails = Site & {
  geometry?: Geometry | null
}
